filebeat/input/journald/pkg/journalctl/reader.go

```diff
@@ -0,0 +1,299 @@
+// journald.
+//
+// File is the journal file to be read, for the system journal use the string
+// `LOCAL_SYSTEM_JOURNAL`.
+//
+// It's the caller's responsibility to call `Close` on the reader to stop
+// the `journalctl` process.
+//
+// If `canceler` is cancelled, the reading goroutine is stopped and subsequent
+// calls to `Next` will return an error.
+func New(
+	logger *logp.Logger,
+	canceler input.Canceler,
+	units []string,
+	syslogIdentifiers []string,
+	transports []string,
+	matchers journalfield.IncludeMatches,
+	mode SeekMode,
+	cursor string,
+	since time.Duration,
+	file string) (*Reader, error) {
+
+	args := []string{"--utc", "--output=json", "--follow"}
+	if file != "" && file != localSystemJournalID {
+		args = append(args, "--file", file)
+	}
+
+	args = handleSeekAndCursor(args, mode, since, cursor)
+
+	for _, u := range units {
+		args = append(args, "--unit", u)
+	}
+
+	for _, i := range syslogIdentifiers {
+		args = append(args, "--identifier", i)
+	}
+
+	for _, m := range matchers.Matches {
+		args = append(args, m.String())
+	}
+
+	for _, m := range transports {
+		args = append(args, fmt.Sprintf("_TRANSPORT=%s", m))
+	}
+
+	logger.Debugf("Journalctl command: journalctl %s", strings.Join(args, " "))
+	cmd := exec.Command("journalctl", args...)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stderr pipe: %w", err)
+	}
+
+	r := Reader{
+		cmd:      cmd,
+		dataChan: make(chan []byte),
+		errChan:  make(chan error),
+		logger:   logger,
+		stdout:   stdout,
+		stderr:   stderr,
+		canceler: canceler,
+	}
+
+	// Goroutine to read errors from stderr
+	r.wg.Add(1)
+	go func() {
+		defer r.logger.Debug("stderr goroutine done")
+		defer r.wg.Done()
+		reader := bufio.NewReader(r.stderr)
+		msgs := []string{}
+		for {
+			line, err := reader.ReadString('\n')
+			if errors.Is(err, io.EOF) {
+				if len(msgs) == 0 {
+					return
+				}
+				errMsg := fmt.Sprintf("Journalctl wrote errors: %s", strings.Join(msgs, "\n"))
+				logger.Errorf(errMsg)
+				r.errChan <- errors.New(errMsg)
+				return
+			}
+			msgs = append(msgs, line)
+		}
+	}()
+
+	// Goroutine to read events from stdout
+	r.wg.Add(1)
+	go func() {
+		defer r.logger.Debug("stdout goroutine done")
+		defer r.wg.Done()
+		reader := bufio.NewReader(r.stdout)
+		for {
+			data, err := reader.ReadBytes('\n')
+			if errors.Is(err, io.EOF) {
+				close(r.dataChan)
+				return
+			}
+
+			select {
+			case <-r.canceler.Done():
+				return
+			case r.dataChan <- data:
+			}
+		}
+	}()
+
+	if err := cmd.Start(); err != nil {
+		return &Reader{}, fmt.Errorf("cannot start journalctl: %w", err)
+	}
+
+	go func() {
+		if err := cmd.Wait(); err != nil {
+			r.logger.Errorf("journalclt exited with an error, exit code %d ", cmd.ProcessState.ExitCode())
```

Comment:
This should set the input status to degraded until the state gets cleared and we manage to restart journalctl

Reply:
Same with the start error above.

Reply:
agreed; if journald fails, it should trickle up to `Next` somehow as well, perhaps?

Reply:
Same thing about the other status reporting comments, we can do this later as an improvement for the input instead of on this PR. Issue to track status reporting: https://github.com/elastic/beats/issues/40143